caneda-help/docs/man/canedafilter.1
.TH CanedaFilter "1" "April 2005" "Debian/GNU Linux" "User Commands"
.SH NAME
CanedaFilter \- A filter synthesis application.
.SH SYNOPSIS
.B canedafilter
[\fIOPTION\fR] ...
.SH DESCRIPTION

\fBCaneda\fR is an integrated circuit simulator which means you will be
able to setup a circuit with a graphical user interface (GUI) and
simulate the large-signal, small-signal and noise behaviour of the
circuit.  After that simulation has finished you will be able to
present the simulation results on a presentation page or window.

The software aims to support all kinds of circuit simulation types,
e.g. DC, AC, S-parameter, harmonic balance analysis, noise analysis,
etc.

\fBCanedaFilter\fR is the filter synthesis tool used by Caneda.  By use of
an input dialog the user can create a filter which is then copied into
the system-wide clipboard.  In \fBCaneda\fR the user opens an empty
schematic and presses CTRL-V (paste from clipboard). The filter
schematic is now inserted and can be simulated.

Available filter types are: Butterworth, Bessel, Chebyshev, Cauer.

.SH AVAILABILITY
The latest version of Caneda can always be obtained from
\fBwww.caneda.tuxfamily.org\fR
.SH "REPORTING BUGS"
Known bugs are documented within the BUGS file.  Report bugs to
<caneda-bugs@lists.tuxfamily.org>.
.SH COPYRIGHT
Copyright \(co 2005 Michael Margraf <michael.margraf@alumni.tu-berlin.de>
.PP
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH AUTHORS
Written by Toyoyuki Ishikawa <ishikawa@conceptkeeper.net> and Michael
Margraf <michael.margraf@alumni.tu-berlin.de>.


ChangeLog
2010-03-25 Renamed project to Caneda
	* Ported all clases and sources to new project name.

2007-10-10 Bastien Roucaries initiated api documentation.
	* qucs/node.h, qucs/node.cpp, qucs/wire.h, qucs/wire.cpp,
	qucs/wireline.h, qucs/item.h, qucs/item.cpp: Add doxygen comments.

2007-03-05  Martín Ribelotta and Lisandro Damián Nicanor Pérez Meyer
	* Ported qucs-edit to qt4
	* Ported qucs-lib to qt4

2006-09-11  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac:  Don't link using "-s" on MacOSX since it drops lazy
	dynamic symbol information.  Also checking for {moc|uic|lrelease|
	lupdate}-qt4 programs.  Added some more paths for Qt headers and
	libraries.

2006-08-08  Stefan Jahn  <stefan@lkcc.org>

	* qucs-attenuator/qucsattenuator.cpp (slotSetText_Zin):
	Synchronized Zout and Zin when specifying a bridged-tee.

	* qucs-attenuator/attenuatorfunc.cpp (Calc): Fixed computation
	routine for bridged-tee topology.

2006-05-29  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.9 and bumped up to version
	0.0.10.

2006-05-14  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac:  Added -s (strip) to linker flags when no
	debugging enabled.
	Put MinGW32 checks in place, thus it is now possible to cross-
	compile the qucs package using a GNU/Linux box.  Had to separate
	LDFLAGS and LIBS because linker command line order matters (-lqt-mt
	-lqtmain *must* be at the end of the command line), otherwise
	linking fails.  Changed rules in appropriate Makefile.am's.
	Additionally checking for the 'ar' tool.

2006-02-16  Stefan Jahn  <stefan@lkcc.org>

	* qucs-transcalc/main.cpp, qucs-lib/main.cpp,
	qucs-help/main.cpp, qucs-filter/main.cpp, qucs-edit/main.cpp
	(loadSettings): Try loading items from main qucsrc even if
	application specific resource file does not exist.

2006-02-10  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac (paths): Added $QTDIR/lib64 to the library path
	to look Qt libraries for.  For 64-bit GNU/Linux's.

2006-01-25  Stefan Jahn  <stefan@lkcc.org>

	* qucs-filter/qf_cauer.cpp: Fixed band stop filters.  Now
	correctly working and exported into the clipboard.

2006-01-21  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.8 and bumped up to version
	0.0.9.

2006-01-18  Stefan Jahn  <stefan@lkcc.org>

	* qucs-filter/qf_filter.cpp: Lowpass/Highpass/Bandpass of
	Cauer filter now quite like working.

	* qucs-filter/qf_cauer.cpp: Fixed bug in bandpass/bandstop
	generation.

2006-01-16  Stefan Jahn  <stefan@lkcc.org>

	* qucs-filter/qf_filter.cpp: Preliminary support for Cauer
	filters.  Many bugs here...

2006-01-14  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Checking for Qt programs also in the
	$QTDIR/bin path.

2006-01-07  Stefan Jahn  <stefan@lkcc.org>

	* qucs-filter/qucsfilter.cpp (calculateFilter): Built in
	the cauer filter code by Vincent Habchi.

2005-11-15  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* show cursor position in qucs-edit
	* extend Bessel coefficient table to order 19

2005-11-08  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* fixed small bug in library search

2005-10-18  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Fixed overwriting CXXFLAGS and CFLAGS
	environment variables via 'configure' command line.

2005-10-13  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* add new libraries: Bridges, LEDs

2005-08-19  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* add new components models to library, thanks to
	Leandro D'Archivio

2005-07-23  Stefan Jahn  <stefan@lkcc.org>

	* qucs-lib/qucslib.1: Added manpage for the QucsLib library
	manager program.

	* configure.ac: Finally fixed freebsd[456]* detection.

2005-07-22  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.7 and bumped up to version
	0.0.8.

2005-05-25  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.6 and bumped up to version
	0.0.7.

2005-05-11  Stefan Jahn  <stefan@lkcc.org>

	* qucs*.1: Updated manpages.

2005-03-01  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.5 and bumped up to version
	0.0.6.

2005-02-27  Stefan Jahn  <stefan@lkcc.org>

	* qucs-help/main.cpp, qucs-edit/main.cpp (main): Made the
	QucsHelp and QucsEdit applications relocatable using the
	environment variable QUCSDIR.

2005-02-23  Stefan Jahn  <stefan@lkcc.org>

	* THANKS: Corrected Toyoyuki Ishikawa's name at all locations
	in the project.

2005-02-21  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: LDFLAGS additions for FreeBSD users.

2005-02-18  Raimund 'Raimi' Jacob  <raimi@lkcc.org>

	* configure.ac: Add another paths to look for Qt header/library in.
	Required for FreeBSD. Use gmake there.

2004-11-24  Raimund 'Raimi' Jacob  <raimi@lkcc.org>

	* autogen.sh: Make aclocal and automake program be overridable
	with environment variable. Need this so i can easily build on
	debian woody, I hope nobody takes offense.
	PS: While Stefan is away, I will commit his stuff.

2004-11-23  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* added icon to QucsEdit and QucsHelp

2004-11-11  Stefan Jahn  <stefan@lkcc.org>

	* qucs-edit/qucsedit.cpp (slotAbout): Added proper copyright
	notice to the 'About' dialog.

2004-10-23  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.4 and bumped up to version
	0.0.5.

2004-09-30  Stefan Jahn  <stefan@lkcc.org>

	* qucs-edit/qucsedit.1: Added simple man page for qucsedit.

	* qucs-help/qucshelp.1: Added simple man page for qucshelp.

	* qucs-help/main.cpp (main): Trying to obtain the correct
	language help directory by splitting e.g. 'es_MX' to use the
	'es' directory.

2004-09-15  Stefan Jahn  <stefan@lkcc.org>

	* qucs-help/main.cpp (main): Using locale() string to check
	whether the documentation directory in the specific language
	is available.

	* configure.ac: Made the DOCDIR variable for choosing a
	specific language.

2004-09-04  Stefan Jahn  <stefan@lkcc.org>


caneda-filter/canedafilter.cpp
/***************************************************************************
                               canedafilter.cpp
                              ----------------
    begin                : Wed Mar 02 2005
    copyright            : (C) 2005 by Michael Margraf
    email                : michael.margraf@alumni.tu-berlin.de
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <math.h>
#include <stdlib.h>
#include <string>

#include <QtGui/QMenuBar>
#include <QtGui/QMenu>
#include <QtGui/QMessageBox>
#include <QtGui/QLayout>
#include <QtGui/QLabel>
#include <QtGui/QPushButton>
#include <QtGui/QLineEdit>
#include <QtGui/QComboBox>
#include <QtGui/QIntValidator>
#include <QtGui/QDoubleValidator>
#include <QtCore/QTimer>
#include <QtGui/QClipboard>
#include <QtGui/QApplication>
#include <QtGui/QPixmap>
#include <QtGui/QGridLayout>

#include "lc_filter.h"
#include "qf_poly.h"
#include "qf_filter.h"
#include "qf_cauer.h"
#include "canedafilter.h"
#include "helpdialog.h"
//There are some bugs with QFontMetrics::width() which causes some display errors
// So I have used some magic numbers to set minimum width.
CanedaFilter::CanedaFilter()
{
  // set application icon
  setWindowIcon (QPixmap(CanedaSettings.BitmapDir + "big.caneda.xpm"));
  setWindowTitle("Caneda Filter " PACKAGE_VERSION);


  // --------  create menubar  -------------------
  QMenuBar *bar = new QMenuBar();
  QMenu *fileMenu = bar->addMenu(tr("&File"));
  bar->addSeparator();
  QMenu *helpMenu = bar->addMenu(tr("&Help"));
  bar->setFixedHeight(bar->sizeHint().height());

  fileMenu->addAction(tr("E&xit"), qApp, SLOT(quit()), Qt::CTRL+Qt::Key_Q);

  helpMenu->addAction(tr("Help..."), this, SLOT(slotHelpIntro()), Qt::Key_F1);
  helpMenu->addSeparator();
  helpMenu->addAction(tr("&About CanedaFilter..."), this, SLOT(slotHelpAbout()), 0);
  helpMenu->addAction(tr("About Qt..."), this, SLOT(slotHelpAboutQt()), 0);


  // -------  create main windows widgets --------
  QVBoxLayout *mainLayout = new QVBoxLayout(this);
  mainLayout->setMargin(0);
  mainLayout->addWidget(bar);
  QGridLayout *gbox = new QGridLayout();
  mainLayout->addLayout(gbox);
  
  QLabel *Label1 = new QLabel(tr("Filter type:"));
  gbox->addWidget(Label1, 0,0);
  ComboType = new QComboBox();
  ComboType->addItem("Bessel");
  ComboType->addItem("Butterworth");
  ComboType->addItem("Chebyshev");
  ComboType->addItem("Cauer");
  gbox->addWidget(ComboType, 0,1);
  gbox->setColumnMinimumWidth(1,ComboType->sizeHint().width()+20);
  connect(ComboType, SIGNAL(activated(int)), this, SLOT(slotTypeChanged(int)));

  QLabel *Label2 = new QLabel(tr("Filter class:"), this);
  gbox->addWidget(Label2, 1,0);
  ComboClass = new QComboBox();
  ComboClass->addItem(tr("Low pass"));
  ComboClass->addItem(tr("High pass"));
  ComboClass->addItem(tr("Band pass"));
  ComboClass->addItem(tr("Band stop"));
  gbox->addWidget(ComboClass, 1,1);
  connect(ComboClass, SIGNAL(activated(int)), this, SLOT(slotClassChanged(int)));

  QIntValidator *IntVal = new QIntValidator(1, 200, this);
  QDoubleValidator *DoubleVal = new QDoubleValidator(this);

  LabelOrder = new QLabel(tr("Order:"));
  gbox->addWidget(LabelOrder, 2,0);
  EditOrder = new QLineEdit("3");
  EditOrder->setValidator(IntVal);
  gbox->addWidget(EditOrder, 2,1);

  LabelStart = new QLabel(tr("Corner frequency:"));
  gbox->addWidget(LabelStart, 3,0);
  EditCorner = new QLineEdit("1");
  EditCorner->setValidator(DoubleVal);
  gbox->addWidget(EditCorner, 3,1);
  ComboCorner = new QComboBox();
  ComboCorner->addItem("Hz");
  ComboCorner->addItem("kHz");
  ComboCorner->addItem("MHz");
  ComboCorner->addItem("GHz");
  ComboCorner->setCurrentIndex(3);
  gbox->addWidget(ComboCorner, 3,2);

  LabelStop = new QLabel(tr("Stop frequency:"));
  gbox->addWidget(LabelStop, 4,0);
  EditStop = new QLineEdit("2");
  EditStop->setValidator(DoubleVal);
  gbox->addWidget(EditStop, 4,1);
  ComboStop = new QComboBox();
  ComboStop->addItem("Hz");
  ComboStop->addItem("kHz");
  ComboStop->addItem("MHz");
  ComboStop->addItem("GHz");
  ComboStop->setCurrentIndex(3);
  gbox->addWidget(ComboStop, 4,2);
  gbox->setColumnMinimumWidth(2,ComboStop->sizeHint().width()+20);

  LabelBandStop = new QLabel(tr("Stop band frequency:"));
  gbox->addWidget(LabelBandStop, 5,0);
  EditBandStop = new QLineEdit("3");
  EditBandStop->setValidator(DoubleVal);
  gbox->addWidget(EditBandStop, 5,1);
  ComboBandStop = new QComboBox();
  ComboBandStop->addItem("Hz");
  ComboBandStop->addItem("kHz");
  ComboBandStop->addItem("MHz");
  ComboBandStop->addItem("GHz");
  ComboBandStop->setCurrentIndex(3);
  gbox->addWidget(ComboBandStop, 5,2);

  LabelRipple = new QLabel(tr("Pass band ripple:"));
  gbox->addWidget(LabelRipple, 6,0);
  EditRipple = new QLineEdit("1");
  EditRipple->setValidator(DoubleVal);
  gbox->addWidget(EditRipple, 6,1);
  LabelRipple_dB = new QLabel("dB");
  gbox->addWidget(LabelRipple_dB, 6,2);

  LabelAtten = new QLabel(tr("Stop band attenuation:"));
  gbox->addWidget(LabelAtten, 7,0);
  gbox->setColumnMinimumWidth(0,LabelAtten->sizeHint().width()+40);
  EditAtten = new QLineEdit("20");
  EditAtten->setValidator(DoubleVal);
  gbox->addWidget(EditAtten, 7,1);
  LabelAtten_dB = new QLabel("dB");
  gbox->addWidget(LabelAtten_dB, 7,2);

  QLabel *Label9 = new QLabel(tr("Impedance:"));
  gbox->addWidget(Label9, 8,0);
  EditImpedance = new QLineEdit("50");
  EditImpedance->setValidator(DoubleVal);
  gbox->addWidget(EditImpedance, 8,1);
  QLabel *Label10 = new QLabel("Ohm");
  gbox->addWidget(Label10, 8,2);


  QPushButton *ButtonGo =
               new QPushButton(tr("Calculate and put into Clipboard"));
  connect(ButtonGo, SIGNAL(clicked()), this, SLOT(slotCalculate()));
  mainLayout->addWidget(ButtonGo);

  LabelResult = new QLabel();
  ResultState = 100;
  slotShowResult();
  LabelResult->setAlignment(Qt::AlignHCenter);
  //LabelResult->setFixedHeight(LabelResult->sizeHint().height());
  mainLayout->addWidget(LabelResult);
  
  // -------  finally set initial state  --------
  slotTypeChanged(0);
  slotClassChanged(0);
}

CanedaFilter::~CanedaFilter()
{
}

// ************************************************************
void CanedaFilter::slotHelpAbout()
{
  QMessageBox::about(this, tr("About..."),
    "CanedaFilter Version " PACKAGE_VERSION+
    tr("\nFilter synthesis program\n")+
    tr("Copyright (C) 2005, 2006 by")+
    "\nVincent Habchi, Toyoyuki Ishikawa,\n"
    "Michael Margraf, Stefan Jahn\n"
    "\nThis is free software; see the source for copying conditions."
    "\nThere is NO warranty; not even for MERCHANTABILITY or "
    "\nFITNESS FOR A PARTICULAR PURPOSE.\n\n");
}

// ************************************************************
void CanedaFilter::slotHelpAboutQt()
{
  QMessageBox::aboutQt(this, tr("About Qt"));
}

// ************************************************************
void CanedaFilter::slotHelpIntro()
{
  HelpDialog *d = new HelpDialog(this);
  d->show();
}

// ************************************************************
void CanedaFilter::setError(const QString& Message)
{
  LabelResult->setText(tr("Result:") + "<font color=\"#FF0000\"><b>  " +
                       tr("Error") + "</b></font>");
  QMessageBox::critical(this, tr("Error"), Message);
}

// ************************************************************
QString * CanedaFilter::calculateFilter(struct tFilter * Filter)
{
  QString * s = NULL;

  if (Filter->Type == TYPE_CAUER) {
    qf_cauer * F = NULL;
    double amin, amax, fc, fs, bw, r;
    fc = Filter->Frequency;
    amin = Filter->Ripple;
    fs = Filter->Frequency3;
    r = Filter->Impedance;
    amax = Filter->Attenuation;
    bw = Filter->Frequency2 - fc;

    switch (Filter->Class) {
    case CLASS_LOWPASS:
      F = new qf_cauer (amin, amax, fc, fs, r, 0, LOWPASS);
      break;
    case CLASS_HIGHPASS:
      F = new qf_cauer (amin, amax, fc, fs, r, 0, HIGHPASS);
      break;
    case CLASS_BANDPASS:
      F = new qf_cauer (amin, amax, fc + bw / 2, fs, r, bw, BANDPASS);
      break;
    case CLASS_BANDSTOP:
      F = new qf_cauer (amin, amax, fc + bw / 2, fs, r, bw, BANDSTOP);
      break;
    }
    if (F) {
      //F->dump();
      EditOrder->setText(QString::number(F->order()));
      s = new QString(F->to_caneda().c_str());
      delete F;
    }
    else {
      s = NULL;
    }
  }
  else {
    s = LC_Filter::createSchematic(Filter);
  }
  return s;
}

// ************************************************************
void CanedaFilter::slotCalculate()
{
  // get numerical values from input widgets
  double CornerFreq   = EditCorner->text().toDouble();
  double StopFreq     = EditStop->text().toDouble();
  double BandStopFreq = EditBandStop->text().toDouble();

  // add exponent
  CornerFreq   *= pow(10, double(3*ComboCorner->currentIndex()));
  StopFreq     *= pow(10, double(3*ComboStop->currentIndex()));
  BandStopFreq *= pow(10, double(3*ComboBandStop->currentIndex()));

  tFilter Filter;
  Filter.Type = ComboType->currentIndex();
  Filter.Class = ComboClass->currentIndex();
  Filter.Order = EditOrder->text().toInt();
  Filter.Ripple = EditRipple->text().toDouble();
  Filter.Attenuation = EditAtten->text().toDouble();
  Filter.Impedance = EditImpedance->text().toDouble();
  Filter.Frequency = CornerFreq;
  Filter.Frequency2 = StopFreq;
  Filter.Frequency3 = BandStopFreq;

  if(EditStop->isEnabled())
    if(Filter.Frequency >= Filter.Frequency2) {
      setError(tr("Stop frequency must be greater than start frequency."));
      return;
    }

  if(EditOrder->isEnabled()) {
    if (Filter.Order < 2) {
      setError(tr("Filter order must not be less than two."));
      return;
    }
    if(Filter.Order > 19) if(Filter.Type == TYPE_BESSEL) {
      setError(tr("Bessel filter order must not be greater than 19."));
      return;
    }
  }

  QString * s = calculateFilter(&Filter);
  if(!s) return;

  // put resulting filter schematic into clipboard
  QClipboard *cb = QApplication::clipboard();
  cb->setText(*s);
  delete s;

  // show result for some time
  ResultState = 0;
  LabelResult->setText(tr("Result:") + "<font color=\"#008000\"><b>  " +
                       tr("Successful") + "</b></font>");
  QTimer::singleShot(500, this, SLOT(slotShowResult()));
}

// ************************************************************
void CanedaFilter::slotShowResult()
{
  if(ResultState > 5) {
    LabelResult->setText(tr("Result: --"));
    return;
  }


  int c;
  ResultState++;
  if(ResultState & 1)  c = 0xFF;
  else c = 0x80;
  QString s = QString("<font color=\"#00%1000\"><b>  ").arg(c, 2, 16);
  LabelResult->setText(tr("Result:") + s + tr("Successful") + "</b></font>");

  c = 500;
  if(ResultState > 5)  c = 3000;
  QTimer::singleShot(c, this, SLOT(slotShowResult()));
}

// ************************************************************
void CanedaFilter::slotTypeChanged(int index)
{
  switch(index) {
    case TYPE_BESSEL:
    case TYPE_BUTTERWORTH:
	LabelRipple->setEnabled(false);
	EditRipple->setEnabled(false);
	LabelRipple_dB->setEnabled(false);
	break;
    case TYPE_CHEBYSHEV:
    case TYPE_CAUER:
	LabelRipple->setEnabled(true);
	EditRipple->setEnabled(true);
	LabelRipple_dB->setEnabled(true);
	break;
  }
  if (index == TYPE_CAUER) {
    LabelOrder->setEnabled(false);
    EditOrder->setEnabled(false);
    LabelAtten->setEnabled(true);
    EditAtten->setEnabled(true);
    LabelAtten_dB->setEnabled(true);
    LabelBandStop->setEnabled(true);
    EditBandStop->setEnabled(true);
    ComboBandStop->setEnabled(true);
  }
  else {
    LabelOrder->setEnabled(true);
    EditOrder->setEnabled(true);
    LabelAtten->setEnabled(false);
    EditAtten->setEnabled(false);
    LabelAtten_dB->setEnabled(false);
    LabelBandStop->setEnabled(false);
    EditBandStop->setEnabled(false);
    ComboBandStop->setEnabled(false);
  }
}

// ************************************************************
void CanedaFilter::slotClassChanged(int index)
{
  switch(index) {
    case CLASS_LOWPASS:
    case CLASS_HIGHPASS:
      LabelStop->setEnabled(false);
      EditStop->setEnabled(false);
      ComboStop->setEnabled(false);
      LabelStart->setText(tr("Corner frequency:"));
      break;
    case CLASS_BANDPASS:
    case CLASS_BANDSTOP:
      LabelStop->setEnabled(true);
      EditStop->setEnabled(true);
      ComboStop->setEnabled(true);
      LabelStart->setText(tr("Start frequency:"));
      break;
  }
  if (index == CLASS_BANDPASS) {
    LabelBandStop->setText(tr("Stop band frequency:"));
    LabelRipple->setText(tr("Pass band ripple:"));
  }
  else if (index == CLASS_BANDSTOP) {
    LabelBandStop->setText(tr("Pass band frequency:"));
    LabelRipple->setText(tr("Pass band attenuation:"));
  }
}
